src/iface-opc-ua.c

```diff
@@ -0,0 +1,347 @@
+/**
+ * This file is part of the KNOT Project
+ *
+ * Copyright (c) 2020, CESAR. All rights reserved.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ */
+
+
+#include <errno.h>
+#include <stdbool.h>
+#include <string.h>
+#include <knot/knot_protocol.h>
+#include <ell/ell.h>
+#include <knot/knot_types.h>
+
+#include <open62541/client.h>
+#include <open62541/client_config_default.h>
+#include <open62541/client_highlevel.h>
+#include <open62541/client_subscriptions.h>
+#include <open62541/plugin/log_stdout.h>
+
+#include "conf-device.h"
+#include "iface-opc-ua.h"
+
+#define RECONNECT_TIMEOUT 5
+
+#define STRING_EMPTY 0
+#define NODEIDTYPE_NUMERIC    "NUMERIC"
+#define NODEIDTYPE_STRING     "STRING"
+#define NODEIDTYPE_GUID       "GUID"
+#define NODEIDTYPE_BYTESTRING "BYTESTRING"
+
+static struct l_timeout *connect_to;
+static struct l_io *opc_ua_io;
+static iface_opc_ua_connected_cb_t conn_cb;
+static iface_opc_ua_disconnected_cb_t disconn_cb;
+struct knot_thing thing_opc_ua;
+static UA_Client *opc_ua_client;
+
+
+union opc_ua_types {
+	float val_float;
+	uint8_t val_bool;
+	int8_t val_i8;
+	uint8_t val_u8;
+	int16_t val_i16;
+	uint16_t val_u16;
+	int32_t val_i32;
+	uint32_t val_u32;
+	int64_t val_i64;
+	uint64_t val_u64;
+};
+
+static UA_ByteString load_cert_file(const char *const path)
+{
+	UA_ByteString fileContents = UA_STRING_NULL;
+	size_t read;
+
+	FILE *fp = fopen(path, "rb");
+
+	if (!fp)
+		return fileContents;
+
+	fseek(fp, 0, SEEK_END);
+	fileContents.length = (size_t)ftell(fp);
+	fileContents.data =
+		(UA_Byte *)UA_malloc(fileContents.length * sizeof(UA_Byte));
+	if (fileContents.data) {
+		fseek(fp, 0, SEEK_SET);
+		read = fread(fileContents.data, sizeof(UA_Byte),
+			     fileContents.length, fp);
+	if (read != fileContents.length)
+		UA_ByteString_clear(&fileContents);
+	} else {
+		fileContents.length = STRING_EMPTY;
+	}
+	fclose(fp);
+
+	return fileContents;
+}
+
+static UA_StatusCode get_node_id(UA_NodeId *parent, int namespace_index,
+			      char *identifier_type,
+			      char *identifier)
+{
+	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+
+	UA_NodeId_init(parent);
+
+	if (!strcmp(identifier_type, NODEIDTYPE_STRING)) {
+		*parent = UA_NODEID_STRING(namespace_index, identifier);
+	} else if (!strcmp(identifier_type, NODEIDTYPE_NUMERIC)) {
+		uint32_t num;
+
+		num = (uint32_t)identifier[0] << 24 |
+		      (uint32_t)identifier[1] << 16 |
+		      (uint32_t)identifier[2] << 8  |
+		      (uint32_t)identifier[3];
```

Comment:
What do you think about extracting this snippet between lines 104 and 107 to a function? We can pass the "identifier" as an argument of the function and receive the return of the operation.